### ISSUE TITLE
borg prune deletes too much

### ISSUE THREAD

**Issue by kmille** [borgbackup/borg#4261]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

A weird behaviour.

## System information. For client/server mode post info for both machines.

Debian Stretch 9.6 on the Client (I do not have any information about the server atm).

#### Your borg version (borg -V).
borg 1.1.8

#### Hardware / network configuration, and filesystems used.
ext4, backuping just one sqlite.db file

#### How much data is handled by borg?
less than 1 MB 

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

see below. It's reproducible:

```
Hey,

our borg prunes backups it should not delete. 


Our backup script:

#!/bin/sh 
 
export HOSTNAME=$(hostname -f) 
export REMOTE='domain.de' 
 
export REPOSITORY=cccda@$REMOTE:$HOSTNAME 
export DATE=`date +%Y-%m-%d_%H:%M` 
export BORG_RSH='ssh -i /root/.ssh/borg_backup_prepaid' 
 
 
borg create                                          \ 
    --stats                                          \ 
    --verbose                                        \
    $REPOSITORY::$DATE                               \
    /opt/strichliste2_backend/var/app.db


borg prune $REPOSITORY                   \
  --verbose                              \
  --list                                 \
  --stats                                \
  --keep-daily    7                      \
  --keep-weekly   4                      \
  --keep-monthly  6


Run it: creates a backup. But it deletes the last one even it's the same and
that should not be deleted because of the --same-day parameter

root@strichliste2:~# ./strichliste-backup.sh 
Creating archive at "cccda@domain.de:strichliste2.w17.io::2019-01-21_16:45"
------------------------------------------------------------------------------
Archive name: 2019-01-21_16:45
Archive fingerprint: ab922288157ab6b8303e55cf21df9a8f2834b5a77eebdefdd6afe8095d90c785
Time (start): Mon, 2019-01-21 16:45:27
Time (end):   Mon, 2019-01-21 16:45:27
Duration: 0.05 seconds
Number of files: 1
Utilization of max. archive size: 0%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:               58.09 kB              5.76 kB                496 B
All archives:              116.18 kB             11.51 kB              6.25 kB

                       Unique chunks         Total chunks
Chunk index:                       4                    6
------------------------------------------------------------------------------
Keeping archive: 2019-01-21_16:45                     Mon, 2019-01-21 16:45:27 [ab922288157ab6b8303e55cf21df9a8f2834b5a77eebdefdd6afe8095d90c785]
Pruning archive: 2019-01-21_16:43                     Mon, 2019-01-21 16:43:20 [056cadf5eeea347ecbf2cdf7a876b61f58404b48c79c2ace935bd463c537b2f1] (1/1)
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
Deleted data:              -58.09 kB             -5.76 kB               -496 B
All archives:               58.09 kB              5.76 kB              5.76 kB

                       Unique chunks         Total chunks
Chunk index:                       3                    3
------------------------------------------------------------------------------

We currently have 1 backup in total.


root@strichliste2:~# borg list $REPOSITORY
2019-01-21_16:45                     Mon, 2019-01-21 16:45:27 [ab922288157ab6b8303e55cf21df9a8f2834b5a77eebdefdd6afe8095d90c785]


Fun fact: --keep-last 7 works for us.

Any idea what we are doing wrong?

Thank you
```

**Comment by ThomasWaldmann** [comment]
```
But it deletes the last one even it's the same and
that should not be deleted because of the --same-day parameter
```
Can you rephrase / fix that? I don't understand it.

But, yeah, if it leaves only 1 archive in total, that looks weird.

**Comment by ThomasWaldmann** [comment]
Can you please remove all trailing blanks at the ends of lines from your script?
Especially if they are **after** a `\`.

**Comment by ThomasWaldmann** [comment]
The specific borg prune output posted above does not show misbehaviour.

It just shows how prune reduces from 2 archives to 1 archive, which is according to the rules, which only keep the last daily backup (for 7 days).

So only the fact that only 2 archives are there at the start of the prune run is strange.

**Comment by kmille** [comment]
Hey,
thanks for the fast response. I tested some different things but it still prunes:
- backup a directory and not a file
- write everything in one line:

I forgot 'day'
> But it deletes the last one even it's the same **day** and
> that should not be deleted because of the --same-day parameter

```
#!/bin/sh

export HOSTNAME=$(hostname -f)
export REMOTE='domain.de'

export REPOSITORY=cccda@$REMOTE:$HOSTNAME
export DATE=`date +%Y-%m-%d_%H:%M`
export BORG_RSH='ssh -i /root/.ssh/borg_backup_prepaid'

borg create --stats --verbose $REPOSITORY::$DATE /opt/strichliste2_backend/var/app.db

borg prune $REPOSITORY --verbose --list --stats --keep-daily 7 --keep-weekly 4 --keep-monthly 6
```

**Comment by ThomasWaldmann** [comment]
There is no "--same-day" option in your script.

**Comment by kmille** [comment]
I mean keep-daily sry!

**Comment by ThomasWaldmann** [comment]
General question: is this a new script or did you already run it daily for a longer time (like for some days at least)?

**Comment by kmille** [comment]
We deployed it yesterday. I wil reread the docs and check the bacups again the next days.

**Comment by ThomasWaldmann** [comment]
OK, keep us updated.

**Comment by ThomasWaldmann** [comment]
any news?

**Comment by kmille** [comment]
Still had no time to look for the issue. I will write you next week. sry

**Comment by dragetd** [comment]
In an attempt to clear it up and summarize it:

--keep-daily (or -weeks, -monthly etc.) takes your archives and checks the creation timestamp of each archive. For --keep-daily 7 it will consider the last 7 days **and pick the newest complete backup each day**, leaving you with 7 archives in total for the last 7 days and **removing any other** archive made within those 7 days. It works the same for more coarse granularity. --keep-monthly 2 for example would delete ALL archive from the last two months except the most recent complete one last and this month. Combining these options (--keep-daily 7 --keep-monthly 1) would keep one at max each of the last 7 days and one for the last month.

Maybe what you want is --keep-within 7d. This tells borg to not delete anything within the last 7 days.

**Comment by mgrachten** [comment]
As @dragetd pointed out --keep-daily 7 is supposed to prune all but the last backup of the last 7 days. That's exactly what is happening in the description of the issue. Indeed it is expected that --keep-last 7 prunes neither of the two backups in the repo, since it will keep the 7 most recent backups.

Since the reported behavior in line with the documentation for the pruning options, I think this issue can be closed.

**Comment by kmille** [comment]
You are right. I misunderstood the docs. I added a --keep-within 3d which is fine for our use case. Thanks for your support!

**Comment by mgrachten** [comment]
I agree it is easy to misinterpret the meaning of the options from the short descriptions in the [docs](https://borgbackup.readthedocs.io/en/stable/usage/prune.html), but I think the longer description and examples below that are quite helpful.
